composeApp/src/commonMain/kotlin/org/ooni/probe/ui/settings/category/SettingsCategoryViewModel.kt

```diff
@@ -0,0 +1,72 @@
+package org.ooni.probe.ui.settings.category
+
+import androidx.datastore.preferences.core.Preferences
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.filterIsInstance
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.update
+import org.ooni.probe.data.repositories.PreferenceCategoryKey
+import org.ooni.probe.data.repositories.PreferenceRepository
+import org.ooni.probe.data.repositories.SettingsKey
+import org.ooni.probe.ui.settings.SettingsCategoryItem
+
+class SettingsCategoryViewModel(
+    preferenceManager: PreferenceRepository,
+    goToSettingsForCategory: (PreferenceCategoryKey) -> Unit,
+    onBack: () -> Unit,
+    category: SettingsCategoryItem,
+) : ViewModel() {
+    private val events = MutableSharedFlow<Event>(extraBufferCapacity = 1)
+
+    private val _state = MutableStateFlow(State(preference = null, category = category))
+    val state = _state.asStateFlow()
+
+    init {
+        category.settings?.map { item -> item.key }?.let { preferenceKeys ->
+            preferenceManager.allSettings(preferenceKeys)
+                .onEach { result -> _state.update { it.copy(preference = result) } }
+                .launchIn(viewModelScope)
+        }
+
+        events.filterIsInstance<Event.SettingsCategoryClick>()
+            .onEach { goToSettingsForCategory(it.category) }.launchIn(viewModelScope)
+
+        events.filterIsInstance<Event.CheckedChangeClick>().onEach {
+            (preferenceManager.preferenceKeyFromSettingsKey(it.key) as? Preferences.Key<Boolean>)?.let { key ->
+                preferenceManager.setValueByKey(
+                    key,
+                    it.value,
+                )
+            }
+            _state.update { state ->
+                state.copy(
+                    preference = state.preference?.plus(it.key to it.value),
+                )
+            }
```

Comment:
You shouldn't need to do this. The state should update automatically since you subscribed above to the flow from the `preferencesManager.allSettings`.